### PR TITLE
Support On-Demand Resource

### PR DESCRIPTION
**Pull request by JunyiXie**
https://github.com/CocoaPods/CocoaPods/issues/9606

### CODE REVIEW

**Comment by shirakaba** [comment]
To provide more context: @JunyiXie has written an excellent [blog post](https://junyixie.github.io/2020/03/19/cocoapods-new-feature/) summarising the changes necessary to `CocoaPods/Core` (this PR) and `CocoaPods/Xcodeproj` (https://github.com/CocoaPods/Xcodeproj/pull/742) in order to support On-Demand Resources. It's quite understandable with Google Translate.

I'm looking forward to this PR!

File: lib/cocoapods-core/specification/consumer.rb [hunk 1]
```diff
@@ -110,6 +110,12 @@ def user_target_xcconfig
         merge_values(attr, value_for_attribute(:xcconfig), value_for_attribute(:user_target_xcconfig))
       end
 
+      # @return [Array<String>] A hash where the key represents the
+      #         paths of the on_demand_resources to copy and the values the paths of
+      #         the on_demand_resources that should be copied.
+      #
+      spec_attr_accessor :on_demand_resources
```
Comment (amorde) [inline lib/cocoapods-core/specification/consumer.rb:117]: I think we would want this to be a `Hash` that allows Pod authors to specify the tag (see my comment [here](https://github.com/CocoaPods/Xcodeproj/pull/742#discussion_r415153967))

example:
```ruby
s.on_demand_resources = {
    'Tag1' => ['file1.png', 'file2.png']
}
```

we also may need to provide a way for Pod authors to specify the category of the tag. The category determines how the resource gets fetched. To quote the [docs](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/On_Demand_Resources_Guide/Tagging.html#//apple_ref/doc/uid/TP40015083-CH3-SW1):

> **Initial install tags**. The resources are downloaded at the same time as the app. The size of the resources is included in the total size for the app in the App Store. The tags can be purged when they are not being accessed by at least one NSBundleResourceRequest object.
>
>**Prefetch tag order**. The resources start downloading after the app is installed. The tags will be downloaded in the order in which they are listed in the Prefetched tag order group.
>
> **Dowloaded only on demand**. The tags are downloaded when requested by the app.

**Comment by amorde** [comment]
Thanks for working on this! I think there's some additional questions we need to answer before this can land, but really appreciate you taking the time to investigate this, and @shirakaba thank you for helping facilitate review

File: lib/cocoapods-core/specification/consumer.rb (hunk 1 above)
Comment (JunyiXie) [reply lib/cocoapods-core/specification/consumer.rb:117]: good idea. I will refer to your design for implementation.

Comment (JunyiXie) [reply lib/cocoapods-core/specification/consumer.rb:117]: implementation https://github.com/CocoaPods/Core/pull/622/commits/f240511fef0d7c56128a336aee7c00f91b718658

**Comment by JunyiXie** [comment]
Thanks for suggestions, I modified the implementation of podspec dsl. @amorde

**Comment by amorde** [comment]
Commenting to say I haven't forgotten about this! Have been taking a break, need a bit more time to review. Thank you for your patience 🙏

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,7 +3,10 @@
 ## Master
 
 ##### Enhancements
-
+* Support On-Demand Resource   
+  [JunyiXie](https://github.com/JunyiXie)
+  [#issue_number](https://github.com/CocoaPods/Core/pull/622)
```
Comment (dnkoutso) [inline CHANGELOG.md:8]: please make this 622 instead of #issue_number

File: lib/cocoapods-core/specification/dsl.rb [hunk 3]
```diff
@@ -1370,6 +1390,7 @@ def dependency=(args)
                 :file_patterns => true,
                 :singularize   => true
 
+
```
Comment (dnkoutso) [inline lib/cocoapods-core/specification/dsl.rb:1393]: unneeded?

File: lib/cocoapods-core/specification/consumer.rb [hunk 4]
```diff
@@ -473,6 +479,24 @@ def _prepare_resource_bundles(value)
...
 
+      # Ensures that the file patterns of the on demand resources are contained in
+      # an array.
+      #
+      # @param  [String, Array, Hash] value.
+      #         The value of the attribute as specified by the user.
+      #
+      # @return [Hash] the resources.
+      #
+      def _prepare_on_demand_resources(value)
```
Comment (dnkoutso) [inline lib/cocoapods-core/specification/consumer.rb:490]: can you please add a test for this logic?

**Comment by dnkoutso** [comment]
superseded by https://github.com/CocoaPods/Core/pull/690

**Pull request closed without merging**